src/circuit/gadgets/constraints.rs

```diff
@@ -430,6 +503,443 @@ pub fn must_be_simple_bit(x: &Boolean) -> AllocatedBit {
     }
 }
 
+// This function enforces a  comparison relation between a and b.
+// It receives as input argument `diff`, which must be constrained to be
+// equal to the difference (a - b).
+// The last argument is `op2`, which can be <, <=, >, >=
+pub fn enforce_comparison<F: LurkField, CS: ConstraintSystem<F>>(
+    mut cs: CS,
+    g: &GlobalAllocations<F>,
+    a: &AllocatedNum<F>,
+    b: &AllocatedNum<F>,
+    diff: &AllocatedNum<F>,
+    op2: &AllocatedNum<F>,
+) -> Result<(Boolean, AllocatedPtr<F>, Boolean), SynthesisError> {
```

Comment:
The return value here, combined with how it is used, make this function much less obviously general than the name or placement implies. It feels like a somewhat random splitting of constraints out from the caller. We should at least document the return values, if they are going to arbitrary.

Reply:
Those values are returned in order to avoid duplicated constraints. Indeed it is much less flexible this way. In order to have a gadget for such functionalities we need to accept such kinds of duplications, unless we come up with a way to look up this information.